pyquil/quil.py

```diff
@@ -138,7 +138,7 @@ def instructions(self) -> List[AbstractInstruction]:
 
         return self._synthesized_instructions
 
-    def inst(self, *instructions: AbstractInstruction) -> "Program":
+    def inst(self, *instructions: Union[Sequence[Any], str, 'Program', DefGate, AbstractInstruction]) -> "Program":
```

Comment:
@appleby is this a sufficiently comprehensive set of possible types of `instructions`?

Reply:
Add a `Generator` to the `Union` and I think it's reasonable. Also, per @karalekas' comment elsewhere, this type hint should match the one for the `*instructions` argument to `__init__`, so it probably makes sense to define a type alias. Something like:

```python
InstructionDesignator = Union[
    AbstractInstruction, 
    DefGate, 
    "Program", 
    Sequence[Any], 
    Generator[Any]
]
```

The `Sequence[Any]` and `Generator[Any]` are not exactly correct, but as long as it doesn't cause type checking problems in practice, I think it's a reasonable shorthand.

Reply:
@appleby should I replace `Generator[Any]` with `Iterator[Any]`? `mypy` complains that `Generator` needs 3 argument, that it has to be `Generator[Any, Any, Any]`. Might as well just use `Iterator`.

Reply:
Actually, I changed my mind. Let's do the following, which most accurately reflects the `isinstance` checks that actually appear in the code. We can always come back later and broaden the types to `Sequence` / `Iterator` / `Iterable` / etc., if desired, but we should then update the `isinstance` checks as well.

```python
InstructionDesignator = Union[
    AbstractInstruction, 
    DefGate, 
    "Program",
    str,  # required to be a pyquil program 
    List[Any], 
    Tuple[Any, ...],
    Generator[Any, Any, Any]
]
```

I haven't tested this either, so hopefully this one doesn't have any errors 😬

Reply:
Tested, doesn't have error.